src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -354,22 +356,35 @@ public String toString() {
     @VisibleForTesting
     abstract class AbstractInternalWriter {
 
-        protected final EventStreamWriter<T> pravegaWriter;
+        @Getter
+        private EventStreamWriter<T> pravegaWriter;
 
-        AbstractInternalWriter(ClientFactory clientFactory) {
+        @Getter
+        private TransactionalEventStreamWriter<T> pravegaTxnWriter;
```

Comment:
Since this is AbstractInternalWriter, why not make it real abstract? such as define only one variable "streamWriter" as AutoCloseable, plus keep the switch of txnWriter (better name it as ifTxnWriter).
The subclass can cast the returned streamWriter appropriately based on ifTxnWriter.

The reason of this suggested refactoring is: object existence test as txn writer flag might not be always reliable. Say if the incoming config is invalid then no writer can be created, but there's no handling of such scenario in constructor, then the object test might be misleading.

Reply:
There is no common interface that can be returned and that's the reason for keeping two separate references in the base class. The child class know which one they are operating on and will not get in to the issue that you are referring to.

Reply:
This sounds a bottom up usage instead of top-down design, then the code looks fine to me